File: python/perfect-numbers/perfect_numbers.py
def classify(number):
    if number < 1:
        raise ValueError(r".+")

    i = 1
    aliquot_sum = 0

    while i * i < number:
        if number % i == 0:
            aliquot_sum += i + number // i

        i += 1

    if i * i == number:
        aliquot_sum += i

    if aliquot_sum - number == number:
        return "perfect"
    elif aliquot_sum - number > number:
        return "abundant"
    else:
        return "deficient"

File: haskell/perfect-numbers/src/PerfectNumbers.hs
module PerfectNumbers
  ( classify
  , Classification(..)
  ) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify :: Int -> Maybe Classification
classify n | n < 1          = Nothing
           | aliquotSum < n = Just Deficient
           | aliquotSum > n = Just Abundant
           | otherwise      = Just Perfect
  where aliquotSum = sum (filter ((0 ==) . mod n) [1 .. (n `div` 2)])
